src/nc_ssl.c

```diff
@@ -0,0 +1,296 @@
+            // This means that the socket needs to do a read or write first.
+            // Since the socket is nonblocking, we can just wait until it is done, per the man page.
+            if (NC_OK != block_until_read_or_write(SSL_get_fd(ssl), 2)) {
+                log_error("Failing SSL_write due to error waiting for read or write.");
+                return NC_ERROR;
+            }
+        }
+        else {
+            log_error("Failing SSL_write due to unhandled error.");
+            log_ssl_error_code(code);
+            return -1;
+        }
+    }
+
+    return bytes_written;
+}
+
+ssize_t
+nc_ssl_writev(SSL *ssl, const struct iovec *iov, int iovcnt) {
+    size_t total_bytes = 0;
+    for (int i = 0; i < iovcnt; i++) {
+        total_bytes += iov[i].iov_len;
+    }
+
+    // FIXME: reconsider using this. it's allocated on the stack so could risk a stackoverflow.
+    // It's not clear if SSL_write() does anything different with buf than write()
+    // This seems to be how writev is implemented (and that's what twemproxy uses for non-SSL).
+    char *buf = alloca(total_bytes);
+
+    copy_all_to_buffer(buf, total_bytes, iov, iovcnt);
+
+    if (total_bytes == 0) {
+        // Calling SSL_write with 0 bytes to send causes undefined behavior.
+        // Since there's nothing to send, return success.
+        return 0;
+    }
+
+    // Must retry failed writes here since there's no guarantee that the caller of this
+    // function will call it with the same arguments. However, it must since (from the man page):
+    // When an SSL_write() operation has to be repeated because of SSL_ERROR_WANT_READ or
+    // SSL_ERROR_WANT_WRITE, it must be repeated with the same arguments.
+    return do_ssl_write(ssl, buf, (int)total_bytes);
+}
+
+ssize_t
+nc_ssl_read(SSL *ssl, void *buf, size_t num) {
+    int bytes_read;
+    while ((bytes_read = SSL_read(ssl, buf, (int)num)) <= 0) {
+        int code = SSL_get_error(ssl, bytes_read);
+
+        if (code == SSL_ERROR_WANT_READ || code == SSL_ERROR_WANT_WRITE) {
+            // This means that the socket needs to do a read or write first.
+            // Since the socket is nonblocking, we can just wait until it is done, per the man page.
+            if (NC_OK != block_until_read_or_write(SSL_get_fd(ssl), 2)) {
+                log_error("Failing SSL_read due to error waiting for read or write.");
+                return NC_ERROR;
+            }
+
+        }
+        else if (code == SSL_ERROR_ZERO_RETURN) {
+            // This is the equivalent of read() returning 0, which happens when the socket is closed.
+            // Return 0 to mimic that behavior.
+            return 0;
+        }
+        else {
+            log_error("Failing SSL_read due to unhandled error.");
+            log_ssl_error_code(code);
+            return -1;
+        }
+    }
+
+    return bytes_read;
+}
+
+static inline size_t
+min(size_t a, size_t b) {
```

Comment:
I ran across docs that recommended against that since it's easy to introduce bugs due to side-effects. e.g. `MIN(++a, b)` increments a twice. Don't have a strong feeling though.